### DISCUSSION TITLE
Why MAX_IMAGE_DIMENSION = 16384 ?

### DISCUSSION THREAD

**Discussion by Charltsing** [d2phap/ImageGlass#1917] [General]
Why MAX_IMAGE_DIMENSION = 16384 ?

Are there any restrictions on image size for DitrctN ?
if i set MAX_IMAGE_DIMENSION = 65536 and open an image with a height of 20000 pixels, DXHelper.ToD2D1Bitmap report error

HResult=0x80004005
  Message=Parameter error
  Source=DirectNStandard
  StackTrace:
   在 DirectN.HRESULT.ThrowOnErrorExcept(IEnumerable`1 exceptedValues, Boolean throwOnError)
   在 DirectN.HRESULT.ThrowOnError(Boolean throwOnError)
   在 DirectN.ID2D1DeviceContextExtensions.CreateBitmapFromWicBitmap[T](ID2D1DeviceContext context, IWICBitmapSource source, Nullable`1 properties)
   在 DirectN.ID2D1DeviceContextExtensions.CreateBitmapFromWicBitmap(IComObject`1 context, IComObject`1 source, Nullable`1 properties)
   在 D2Phap.DXHelper.ToD2D1Bitmap(IComObject`1 dc, WicBitmapSource wicSrc, Nullable`1 bitmapProps)

Does DirectN not support large images?
How to open a large image with UseHardwareAcceleration ?

**Comment by d2phap** [comment]
It's the limit of Max Texture Dimension in DirectX:
https://learn.microsoft.com/en-us/windows/win32/direct3d11/overviews-direct3d-11-devices-downlevel-intro?redirectedfrom=MSDN#direct3d-12-feature-support-feature-levels-12_2-through-11_0

For larger size, only read the small region from the source file and load to Direct2D, it's complicated.

**Comment by Charltsing** [comment]
https://stackoverflow.com/questions/16337760/large-images-with-direct2d

For example,

Use ID2D1RenderTarget::CreateSharedBitmap to make a bitmap that can be used by different resources.
Then create a ID2D1BitmapRenderTarget and render the large bitmap into that. (making sure to do BeginDraw, Clear, DrawBitmap, EndDraw). Both the bitmap and the render target can be cached for use by successive calls.
Then copy from that render target into a regular ID2D1Bitmap with the portion that will fit into the texture memory using the ID2D1Bitmap::CopyFromRenderTarget method.
Finally draw that to the real render target, pRT->DrawBitmap



Or, at the appropriate time, resize the image to a size smaller than 16384
